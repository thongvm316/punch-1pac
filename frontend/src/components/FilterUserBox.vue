<template>
  <v-select
    v-model="selectedUser"
    label="email"
    :filterable="false"
    :placeholder="placeholder"
    :options="optionUsers"
    @search="filterUsers"
    @input="updateSelectedUser"
  >
    <template
      slot="option"
      slot-scope="option"
    >
      <div class="tile tile-centered">
        <div class="tile-icon">
          <img
            :src="option.avatar_url"
            class="avatar avatar-md"
          >
        </div>
        <div class="tile-content">
          {{ option.name }} ({{ option.email }})
        </div>
      </div>
    </template>
    <span slot="no-options">{{ $t('filterUserBox.noOptions') }}</span>
  </v-select>
</template>

<script>
import Repositories from '@/repository'
import debounce from 'lodash.debounce'
const vSelect = () => import('vue-select')

export default {
  components: {
    vSelect
  },

  props: {
    queryParams: {
      type: Object,
      required: true
    },
    placeholder: {
      type: String,
      default: ''
    },
    user: {
      type: Object,
      default: null
    }
  },

  data() {
    return {
      optionUsers: [],
      selectedUser: null
    }
  },

  watch: {
    user: function() {
      this.optionUsers = []
      this.selectedUser = this.user
    }
  },

  created() {
    this.search('', false)
  },

  methods: {
    filterUsers: debounce(function(search, loading) {
      loading(true)
      this.search(search, loading)
    }, 350),

    search(search, loading) {
      const params = Object.assign({ name_or_email: search }, this.queryParams)

      Repositories.get('users').getUsers(params)
        .then(response => {
          if (loading) loading(false)
          this.optionUsers = response.data.users
        })
        .catch(error => {
          throw error
        })
    },

    updateSelectedUser() {
      this.$emit('update:user', this.selectedUser)
    }
  }
}
</script>
