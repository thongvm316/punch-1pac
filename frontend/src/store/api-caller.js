import axios from 'axios'

export default function callApi(payload = { params: null, data: null }) {
  const { method, url, params, data } = payload

  return axios({
    method,
    url,
    params,
    data
  })
}
