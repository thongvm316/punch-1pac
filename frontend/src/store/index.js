import Vue from 'vue'
import Vuex from 'vuex'
import createLogger from 'vuex/dist/logger'

import headerAnnouncements from './modules/header-announcements'
import punch from './modules/punch'

import companyUsers from './modules/company-users'
import companyDepartments from './modules/company-departments'

import announcements from './modules/announcements'
import announcement from './modules/announcement'
import userSessions from './modules/user-sessions'
import attendances from './modules/attendances'
import companyAllowedIPs from './modules/company-allowed_ips'

Vue.use(Vuex)

const debug = process.env.NODE_ENV !== 'production'

export default new Vuex.Store({
  modules: {
    headerAnnouncements,
    punch,
    companyUsers,
    companyDepartments,
    announcements,
    announcement,
    userSessions,
    attendances,
    companyAllowedIPs
  },
  strict: debug,
  plugins: debug ? [createLogger()] : []
})
