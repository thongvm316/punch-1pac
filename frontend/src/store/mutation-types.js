// initial-states
export const INITIAL_STATES_SET_USER = 'INITIAL_STATES_SET_USER'
export const INITIAL_STATES_SET_COMPANY = 'INITIAL_STATES_SET_COMPANY'
export const INITIAL_STATES_SET_META = 'INITIAL_STATES_SET_META'
export const INITIAL_STATES_UPDATE_USER = 'INITIAL_STATES_UPDATE_USER'
export const INITIAL_STATES_UPDATE_COMPANY = 'INITIAL_STATES_UPDATE_COMPANY'
export const INITIAL_STATES_SET_USER_ERRORS = 'INITIAL_STATES_SET_USER_ERRORS'
export const INITIAL_STATES_SET_COMPANY_ERRORS = 'INITIAL_STATES_SET_COMPANY_ERRORS'
export const INITIAL_STATES_CLEAR_USER_ERRORS = 'INITIAL_STATES_CLEAR_USER_ERRORS'
export const INITIAL_STATES_CLEAR_COMPANY_ERRORS = 'INITIAL_STATES_CLEAR_COMPANY_ERRORS'
export const INITIAL_STATES_UPDATE_PASSWORD_CHANGED = 'INITIAL_STATES_UPDATE_PASSWORD_CHANGED'

// announcements
export const READ_ANNOUNCEMENT = 'READ_ANNOUNCEMENT'
export const RECEIVE_HEADER_ANNOUNCEMENTS = 'RECEIVE_HEADER_ANNOUNCEMENTS'

// attendances
export const RECEIVE_ATTENDANCES = 'RECEIVE_ATTENDANCES'

// punch
export const PUNCH_IN = 'PUNCH_IN'
export const PUNCH_OUT = 'PUNCH_OUT'
export const PUNCH_INIT_ATTENDANCE = 'PUNCH_INIT_ATTENDANCE'

// company-users
export const FETCH_USERS = 'FETCH_USERS'
export const DELETE_USER = 'DELETE_USER'
export const UPDATE_USER = 'UPDATE_USER'

// company-business-days
export const FETCH_BUSINESS_DAYS = 'FETCH_BUSINESS_DAYS'
export const ADD_BUSINESS_DAY = 'ADD_BUSINESS_DAY'
export const DELETE_BUSINESS_DAY = 'DELETE_BUSINESS_DAY'
export const UPDATE_BUSINESS_DAY = 'UPDATE_BUSINESS_DAY'
export const UPDATE_BUSINESS_DAY_ERRORS = 'UPDATE_BUSINESS_DAY_ERRORS'
export const CLEAR_BUSINESS_DAY_ERRORS = 'CLEAR_BUSINESS_DAY_ERRORS'

// company-allowed_ips
export const FETCH_IPS = 'FETCH_IPS'
export const CREATE_IP = 'CREATE_IPS'
export const UPDATE_IP = 'UPDATE_IPS'
export const DELETE_IP = 'DELETE_IPS'
export const UPDATE_IP_ERRORS = 'UPDATE_REQUEST_ERRORS'
export const CLEAR_IP_ERRORS = 'CLEAR_REQUEST_ERRORS'

// company-holidays
export const FETCH_HOLIDAYS = 'FETCH_HOLIDAYS'
export const CREATE_HOLIDAY = 'CREATE_HOLIDAY'
export const UPDATE_HOLIDAY = 'UPDATE_HOLIDAY'
export const DELETE_HOLIDAY = 'DELETE_HOLIDAY'
export const UPDATE_HOLIDAY_ERRORS = 'UPDATE_HOLIDAY_ERRORS'
export const CLEAR_HOLIDAY_ERRORS = 'CLEAR_HOLIDAY_ERRORS'
export const IMPORT_NATIONAL_HOLIDAYS = 'IMPORT_NATIONAL_HOLIDAYS'

// user-sessions
export const FETCH_SESSIONS = 'FETCH_SESSIONS'
export const DELETE_SESSION = 'DELETE_SESSION'

// user-password
export const UPDATE_USER_PASSWORD_ERRORS = 'UPDATE_USER_PASSWORD_ERRORS'
export const CLEAR_USER_PASSWORD_ERRORS = 'CLEAR_USER_PASSWORD_ERRORS'

// requests
export const RECEIVE_REQUESTS = 'RECEIVE_REQUESTS'
export const ADD_REQUEST = 'ADD_REQUEST'
export const UPDATE_REQUEST = 'UPDATE_REQUEST'
export const DELETE_REQUEST = 'DELETE_REQUEST'
export const UPDATE_REQUEST_ERRORS = 'UPDATE_REQUEST_ERRORS'
export const CLEAR_REQUEST_ERRORS = 'CLEAR_REQUEST_ERRORS'

// group-requests
export const RECEIVE_GROUP_REQUESTS = 'RECEIVE_GROUP_REQUESTS'
export const APPROVE_GROUP_REQUEST = 'APPROVE_GROUP_REQUEST'
export const REJECT_GROUP_REQUEST = 'REJECT_GROUP_REQUEST'
export const REJECT_GROUP_REQUEST_ERRORS = 'REJECT_GROUP_REQUEST_ERRORS'
export const CLEAR_REJECT_GROUP_REQUEST_ERRORS = 'CLEAR_REJECT_GROUP_REQUEST_ERRORS'

// group-attendances
export const RECEIVE_GROUP_ATTENDANCES = 'RECEIVE_GROUP_ATTENDANCES'
export const FETCH_USERS_IN_GROUP_ATTENDANCES = 'FETCH_USERS_IN_GROUP_ATTENDANCES'

// group-report
export const FETCH_GROUP_REPORT = 'FETCH_GROUP_REPORT'

// groups
export const RECEIVE_GROUPS = 'RECEIVE_GROUPS'
export const ADD_GROUP = 'ADD_GROUP'
export const UPDATE_GROUPS_ERRORS = 'UPDATE_GROUPS_ERRORS'
export const CLEAR_GROUPS_ERRORS = 'CLEAR_GROUPS_ERRORS'

// group
export const RECEIVE_GROUP = 'RECEIVE_GROUP'
export const UPDATE_GROUP = 'UPDATE_GROUP'
export const ADD_GROUP_USER = 'ADD_GROUP_USER'
export const REMOVE_GROUP_USER = 'REMOVE_GROUP_USER'
export const UPDATE_GROUP_ERRORS = 'UPDATE_GROUP_ERRORS'
export const CLEAR_GROUP_ERRORS = 'CLEAR_GROUP_ERRORS'
export const FETCH_USERS_NOT_IN_GROUP = 'FETCH_USERS_NOT_IN_GROUP'

// chart
export const FETCH_CHART_DATA = 'FETCH_CHART_DATA'
export const RESET_CHART_DATA = 'RESET_CHART_DATA'

// calendar
export const FETCH_CALENDAR_ATTENDANCES = 'FETCH_CALENDAR_ATTENDANCES'

// flash
export const SET_FLASH_MESSAGE = 'SET_FLASH_MESSAGE'

// notifications
export const FETCH_HEADER_NOTIFICATIONS = 'FETCH_HEADER_NOTIFICATIONS'
export const FETCH_NEW_HEADER_NOTIFICATIONS = 'FETCH_NEW_HEADER_NOTIFICATIONS'
export const FETCH_MORE_HEADER_NOTIFICATIONS = 'FETCH_MORE_HEADER_NOTIFICATIONS'
export const READ_NOTIFICATIONS = 'READ_NOTIFICATIONS'
export const SET_INTERVAL_FETCH_NOTIFICATIONS = 'SET_INTERVAL_FETCH_NOTIFICATIONS'

// activities
export const FETCH_ACTIVITIES = 'FETCH_ACTIVITIES'
export const FETCH_MORE_ACTIVITIES = 'FETCH_MORE_ACTIVITIES'

// status-cards
export const FETCH_STATUS_CARDS = 'FETCH_STATUS_CARDS'
