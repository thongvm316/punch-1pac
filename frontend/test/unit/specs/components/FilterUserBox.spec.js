import { shallowMount } from '@vue/test-utils'

import localVue from '../../supports/local-vue'
import vSelect from 'vue-select'

import store from '@/store'
import i18n from '@/locale'
import FilterUserBox from '@/components/FilterUserBox'

const queryParams = {
  email: 'example@1pac.vn',
  name: 'Tuan',
  avatar_url: '/'
}

const search = jest.fn()

describe('FilterUserBox.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(FilterUserBox, {
      i18n,
      store,
      localVue,
      propsData: {
        queryParams
      },
      methods: {
        search
      }
    })
  })

  afterEach(() => {
    wrapper.vm.$destroy()
  })

  describe('when FilterUserBox was mounted', () => {
    it('should display FilterUserBox component', () => {
      expect(wrapper.isVueInstance()).toBe(true)
      expect(wrapper.exists()).toBe(true)
    })

    it('should validate its props', () => {
      const { placeholder, queryParams, user } = wrapper.vm.$options.props

      expect(queryParams.required).toBeTruthy()
      expect(queryParams.type).toBe(Object)

      expect(placeholder.type).toBe(String)
      expect(user.type).toBe(Object)
    })

    it('should render sub components', () => {
      expect(wrapper.find(vSelect).exists()).toBe(true)
    })

    it('method search should been called', () => {
      expect(search).toHaveBeenCalled()
    })

    describe('when passing props user', () => {
      const user = {
        email: 'example@1pac.vn',
        name: 'Tuan',
        avatar_url: '/'
      }

      beforeEach(() => {
        wrapper.setProps({ user })
      })

      it('selectedUser should have user data from props', () => {
        wrapper.vm.$nextTick(() => {
          expect(wrapper.vm.$data.selectedUser).toEqual(user)
        })
      })

      it('should emit update:user with user data', () => {
        wrapper.vm.$nextTick(() => {
          wrapper.vm.updateSelectedUser()
          expect(wrapper.emitted('update:user')).toBeTruthy()
          expect(wrapper.emitted('update:user')).toHaveLength(1)
          expect(wrapper.emitted('update:user')[0]).toEqual([user])
        })
      })
    })
  })
})
