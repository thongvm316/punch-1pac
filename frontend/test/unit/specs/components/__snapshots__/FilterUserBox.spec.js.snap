// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterUserBox.vue when FilterUserBox was mounted when rendered should display FilterUserBox component 1`] = `<vselect-stub label="email" placeholder="Search box" options="[object Object],[object Object],[object Object]"> <span>No user matching</span></vselect-stub>`;
